### PR TITLE
Histogram performance improvements

### PR DESCRIPTION
**Pull request by basilisk487**

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java [hunk 1]
```diff
@@ -2,53 +2,108 @@
...
- * Binary format is length[2b], {utf8Len[2b], utfSeq}*
+ * Binary format is signature[4b] = 0x54415045, version[4b] = 0x00000001, length[4b], {utf8Len[4b], utfSeq}*
  *
  * @author Tim Schmidt (tim@wavefront.com).
+ * @author Vasily Vorontsov (vasily@wavefront.com)
+ *
  */
 public class TapeStringListConverter implements FileObjectQueue.Converter<List<String>> {
-  private static final TapeStringListConverter INSTANCE = new TapeStringListConverter();
+  private static final Logger logger = LogManager.getLogger(TapeStringListConverter.class.getCanonicalName());
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: just use java.util.logging

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java [hunk 2]
```diff
@@ -2,53 +2,108 @@
...
+  public static TapeStringListConverter getDefaultInstance() {
+    return INSTANCE_DEFAULT;
+  }
+
+  public static TapeStringListConverter getCompressionEnabledInstance() {
+    return INSTANCE_COMPRESSION_ENABLED;
   }
 
   @Override
   public List<String> from(byte[] bytes) throws IOException {
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: This public method is complex enough and should have a javadoc

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java [hunk 3]
```diff
@@ -2,53 +2,108 @@
...
-    List<String> result = new ArrayList<>(count);
-    for (int i = 0; i < count; ++i) {
-      int len = in.getShort();
-      result.add(new String(bytes, in.position(), len));
-      in.position(in.position() + len);
+    try {
+      ByteBuffer in;
+      byte[] uncompressedData;
+      if (bytes.length > 2 && bytes[0] == (byte) 0x4C && bytes[1] == (byte) 0x5A) { // LZ block signature
+        uncompressedData = IOUtils.toByteArray(new LZ4BlockInputStream(new ByteArrayInputStream(bytes)));
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: Should you wrap and assume uncompressed?

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java [hunk 4]
```diff
@@ -2,53 +2,108 @@
...
+        in = ByteBuffer.wrap(uncompressedData);
+      } else {
+        uncompressedData = bytes;
+        in = ByteBuffer.wrap(bytes);
+      }
+      int signature = in.getInt();
+      int version = in.getInt();
+      if (signature != CONTAINER_SIGNATURE || version != CONTAINER_VERSION) {
+        logger.error("Unknown data format retrieved from tape (signature = " + signature + ", version = " + version);
```
Comment (panghy) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: Our error messages have a key and should probably stick to one (at this point we should also have a list of error codes in the README.md)

**Comment by basilisk487** [comment]
@panghy: Surprisingly, Math.asin() actually _was_ the biggest bottleneck and highest CPU consumer in merging points into histograms:
image
Switching to jafama bumps performance from ~80kpps to ~375kpps, and reduces CPU usage for com.tdunning.math.stats.AgentDigest.mergeNewValues() from ~42% to ~5%.

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java (hunk 1 above)
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: Done

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java (hunk 2 above)
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: Done

File: proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java (hunk 4 above)
Comment (basilisk487) [inline proxy/src/main/java/com/wavefront/agent/histogram/tape/TapeStringListConverter.java]: Done. I'll take care of remaining error codes + README.md in a separate PR

**Comment by basilisk487** [comment]
@panghy: GZIP still has quite a bit of overhead, especially considering that all histogram traffic goes through the "receive -> compress -> write to file -> read from file -> decompress -> parse" cycle by default.
Measured throughput (same configuration, identical data shape): 
- GZIP with default compression: 280kpps
- GZIP with level 2 compression: 310kpps
- LZ4: 400kpps
Given the above I believe it's still worth it to stick with LZ4, especially considering that we are already using it for the retry queue anyway, as GZIP was pegging CPU on a 4-core machine when proxy was overloaded and forced to flush to the retry queue

**Comment by panghy** [comment]
Thanks, it's good to know about fastmath.

**Pull request merged**